contact_map/atom_indexer.py

```diff
@@ -0,0 +1,107 @@
+    topology._atoms = indices
+    topology._numAtoms = len(indices)
+    return md.Trajectory(xyz=xyz, topology=topology, time=time,
+                         unitcell_lengths=unitcell_lengths,
+                         unitcell_angles=unitcell_angles)
+
+def residue_query_atom_idxs(sliced_query, atom_idx_to_residue_idx):
+    residue_query_atom_idxs = collections.defaultdict(list)
+    for sliced_idx in sliced_query:
+        residue_idx = atom_idx_to_residue_idx[sliced_idx]
+        residue_query_atom_idxs[residue_idx].append(sliced_idx)
+    return residue_query_atom_idxs
+
+
+class AtomSlicedIndexer(object):
+    """Indexer when using atom slicing.
+    """
+    def __init__(self, topology, real_query, real_haystack, all_atoms):
+        self.all_atoms = all_atoms
+        self.sliced_idx = {
+            real_idx : sliced_idx
+            for sliced_idx, real_idx in enumerate(all_atoms)
+        }
+        self.real_idx = {
+            sliced_idx: real_idx
+            for real_idx, sliced_idx in self.sliced_idx.items()
+        }
+        self.query = set([self.sliced_idx[q] for q in real_query])
+        self.haystack = set([self.sliced_idx[h] for h in real_haystack])
+
+        # atom_idx_to_residue_idx
+        self.real_atom_idx_to_residue_idx = {atom.index: atom.residue.index
+                                             for atom in topology.atoms}
+        self.atom_idx_to_residue_idx = {
+            sliced_idx: self.real_atom_idx_to_residue_idx[real_idx]
+            for sliced_idx, real_idx in enumerate(all_atoms)
+        }
+        self.residue_query_atom_idxs = residue_query_atom_idxs(
+            self.query, self.atom_idx_to_residue_idx
+        )
+
+    def ignore_atom_idx(self, atoms, all_atoms_set):
+        result = set(atom.index for atom in atoms)
+        result &= all_atoms_set
+        result = set(self.sliced_idx[a] for a in result)
+        return result
+
+    def convert_atom_contacts(self, atom_contacts):
+        result =  {frozenset(map(self.real_idx.__getitem__, pair)): value
+                   for pair, value in atom_contacts.items()}
+        return collections.Counter(result)
+
+    def slice_trajectory(self, trajectory):
+        # Prevent (memory) expensive atom slicing if not needed.
+        # This check is also needed here because ContactFrequency slices the
+        # whole trajectory before calling this function.
+        if len(self.all_atoms) < trajectory.topology.n_atoms:
+            sliced = _atom_slice(trajectory, self.all_atoms)
+            sliced_trajectory = _atom_slice(trajectory, self.all_atoms)
+        else:
+            sliced = trajectory
+        return sliced
+
+
+class IdentityIndexer(object):
+    """Indexer when not using atom slicing.
+    """
+    def __init__(self, topology, real_query, real_haystack, all_atoms):
+        self.all_atoms = all_atoms
+        self.topology = topology
+        identity_mapping = {a: a for a in range(topology.n_atoms)}
```

Comment:
As I would classify this as "performant" code, maybe only make dicts for the atoms used to reduce Memory (which might be less than all atoms, even without atom_slice)?
```suggestion
        identity_mapping = {a: a for a in all_atoms)}
```

feel free to ignore

Reply:
I agree, and implemented it this way to start. However, that broke tests here (`KeyError: 0` on line 327):

https://github.com/dwhswenson/contact_map/blob/0f5201bd80be33b0207d687b9bc884e11e4a72e3/contact_map/tests/test_contact_map.py#L324-L331

I tried to change the tests as little as possible to avoid breaking intended behavior. The change to that test that works with the proposed change is:

```python
        # Test sliced indices
        sliced_idx = [0, 1, 2, 3]
        real_idx = [map0b.indexer.real_idx.get(i, None) for i in sliced_idx]
        if map0b._use_atom_slice:
            assert real_idx == [1, 4, 5, 6]
        else:
            assert real_idx == [None, 1, None, None]
```

I find it a little counterintuitive that `indexer.real_idx[0]` doesn't exist.

Can go either way; leaning toward leaving as is because this isn't *that* performance-focused. The only use case where this is a likely performance hit is where you're doing all-atom contacts for many different systems; e.g., proteomics. (Not all atoms? Use slicing to speed you up. Same structure? Use `ContactTrajectory`; this only happens once.)